File: tools/docs-generator/lib/physics/fixture/fixture_prefab_xml_reader.py
import xml.etree.ElementTree as EXml

from lib.physics.filter.filter_prefab_xml_reader import FilterPrefabXmlReader
from lib.physics.fixture.fixture_prefab import FixturePrefab
from lib.physics.shape.shape_prefab_xml_reader import ShapePrefabXmlReader


class FixturePrefabXmlReader(object):
    FIXTURE_ELEMENT = "fixture"
    RESTITUTION_ELEMENT = "restitution"
    DENSITY_ELEMENT = "density"
    FRICTION_ELEMENT = "friction"
    ID_ATTRIBUTE = "id"

    DEFAULT_DENSITY = 0.0
    DEFAULT_FRICTION = 0.0
    DEFAULT_RESTITUTION = 0.0

    def __init__(self, filter_prefab_xml_reader, shape_prefab_xml_reader):
        self.filter_prefab_xml_reader = filter_prefab_xml_reader
        self.shape_prefab_xml_reader = shape_prefab_xml_reader

    def read_prefab_from_string(self, xml_string, is_sensor):
        try:
            element = EXml.fromstring(xml_string)
            filter_prefab = self.filter_prefab_xml_reader.read_prefab_from_string(
                EXml.tostring(element.find(FilterPrefabXmlReader.FILTER_ELEMENT))
            )
            shape_prefab = self.shape_prefab_xml_reader.read_prefab_from_string(
                EXml.tostring(element.find(ShapePrefabXmlReader.SHAPE_ELEMENT))
            )

            restitution = self.DEFAULT_RESTITUTION
            density = self.DEFAULT_DENSITY
            friction = self.DEFAULT_FRICTION

            if element.find(self.RESTITUTION_ELEMENT) is not None:
                restitution = float(element.find(self.RESTITUTION_ELEMENT).text.strip())

            if element.find(self.DENSITY_ELEMENT) is not None:
                density = float(element.find(self.DENSITY_ELEMENT).text.strip())

            if element.find(self.FRICTION_ELEMENT) is not None:
                friction = float(element.find(self.FRICTION_ELEMENT).text.strip())

            return FixturePrefab(filter_prefab, shape_prefab, restitution, density, friction, is_sensor)
        except Exception as e:
            raise FixturePrefabXmlReadException(xml_string, e)


class FixturePrefabXmlReadException(Exception):
    MESSAGE_TEMPLATE = "Cannot read fixture prefab from xml string {}. Cause: {}."

    def __init__(self, xml_string, cause):
        super().__init__(self.MESSAGE_TEMPLATE.format(xml_string, cause))

File: core/src/remove/tanks/game/physics/fixture/FixturePrefabXmlReader.java
package remove.tanks.game.physics.fixture;

import com.badlogic.gdx.utils.XmlReader;
import remove.tanks.game.physics.filter.FilterPrefab;
import remove.tanks.game.physics.filter.FilterPrefabXmlReader;
import remove.tanks.game.physics.shape.ShapePrefab;
import remove.tanks.game.physics.shape.ShapePrefabXmlReader;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Mateusz Długosz
 */
public final class FixturePrefabXmlReader {
    public static final String FIXTURES_ELEMENT = "fixtures";
    public static final String FIXTURE_ELEMENT = "fixture";

    private static final String FRICTION_ELEMENT = "friction";
    private static final String DENSITY_ELEMENT = "density";
    private static final String RESTITUTION_ELEMENT = "restitution";
    private static final String IS_SENSOR_ELEMENT = "isSensor";

    private static final float FRICTION_DEFAULT_VALUE = 0f;
    private static final float DENSITY_DEFAULT_VALUE = 0f;
    private static final float RESTITUTION_DEFAULT_VALUE = 0f;
    private static final boolean IS_SENSOR_DEFAULT_VALUE = false;

    private final ShapePrefabXmlReader shapePrefabXmlReader;
    private final FilterPrefabXmlReader filterPrefabXmlReader;

    public FixturePrefabXmlReader(
            ShapePrefabXmlReader shapePrefabXmlReader,
            FilterPrefabXmlReader filterPrefabXmlReader
    ) {
        this.shapePrefabXmlReader = shapePrefabXmlReader;
        this.filterPrefabXmlReader = filterPrefabXmlReader;
    }

    public FixturePrefab readFixturePrefabAsHitBox(XmlReader.Element element) {
        try {
            return new FixturePrefab(
                    readFilterPrefab(element),
                    readShapePrefab(element),
                    readRestitution(element),
                    readDensity(element),
                    readFriction(element),
                    false
            );
        } catch (Exception e) {
            throw new FixturePrefabXmlReadException(element, e);
        }
    }

    public FixturePrefab readFixturePrefabAsSensor(XmlReader.Element element) {
        try {
            return new FixturePrefab(
                    readFilterPrefab(element),
                    readShapePrefab(element),
                    readRestitution(element),
                    readDensity(element),
                    readFriction(element),
                    false
            );
        } catch (Exception e) {
            throw new FixturePrefabXmlReadException(element, e);
        }
    }

    private FilterPrefab readFilterPrefab(XmlReader.Element element) {
        return filterPrefabXmlReader.readFilterPrefab(
                element.getChildByName(FilterPrefabXmlReader.FILTER_ELEMENT)
        );
    }

    private ShapePrefab readShapePrefab(XmlReader.Element element) {
        return shapePrefabXmlReader.readShapePrefab(
                element.getChildByName(ShapePrefabXmlReader.SHAPE_ELEMENT)
        );
    }

    private float readRestitution(XmlReader.Element element) {
        if (element.getChildByName(RESTITUTION_ELEMENT) == null) {
            return RESTITUTION_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(RESTITUTION_ELEMENT).getText().trim());
    }

    private float readDensity(XmlReader.Element element) {
        if (element.getChildByName(DENSITY_ELEMENT) == null) {
            return DENSITY_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(DENSITY_ELEMENT).getText().trim());
    }

    private float readFriction(XmlReader.Element element) {
        if (element.getChildByName(FRICTION_ELEMENT) == null) {
            return FRICTION_DEFAULT_VALUE;
        }
        return Float.valueOf(element.getChildByName(FRICTION_ELEMENT).getText().trim());
    }
}
